File: backend/app/domain/models/achievement.py
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from uuid import UUID


class AchievementType(str, Enum):
    STREAK = "streak"
    TOTAL_HABITS = "total_habits"
    LEVEL = "level"
    POINTS = "points"
    PERFECT_WEEK = "perfect_week"


@dataclass
class Achievement:
    id: UUID
    name: str
    description: str
    icon: str
    achievement_type: AchievementType
    condition_value: int
    reward_points: int
    is_active: bool

    def check_condition(self, user_value: int) -> bool:
        return user_value >= self.condition_value


@dataclass
class UserAchievement:
    id: UUID
    user_id: UUID
    achievement_id: UUID
    earned_at: datetime
    notified: bool

    def mark_as_notified(self) -> None:
        self.notified = True

File: frontend/lib/features/gamification/domain/entities/achievement.dart
import 'package:equatable/equatable.dart';

enum AchievementType {
  streak,
  totalHabits,
  level,
  points,
  perfectWeek;

  String get displayName {
    switch (this) {
      case AchievementType.streak:
        return 'Серия';
      case AchievementType.totalHabits:
        return 'Всего привычек';
      case AchievementType.level:
        return 'Уровень';
      case AchievementType.points:
        return 'Баллы';
      case AchievementType.perfectWeek:
        return 'Идеальная неделя';
    }
  }
}

class Achievement extends Equatable {
  const Achievement({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.type,
    required this.conditionValue,
    required this.rewardPoints,
    required this.isEarned,
    required this.progress,
    required this.progressPercent,
    this.earnedAt,
  });

  final String id;

  final String name;

  final String description;

  final String icon;

  final AchievementType type;

  final int conditionValue;

  final int rewardPoints;

  final bool isEarned;

  final int progress;

  final double progressPercent;

  final DateTime? earnedAt;

  bool get isInProgress => !isEarned && progress > 0;

  bool get isLocked => !isEarned && progress == 0;

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        icon,
        type,
        conditionValue,
        rewardPoints,
        isEarned,
        progress,
        progressPercent,
        earnedAt,
      ];

  @override
  String toString() {
    return 'Achievement(id: $id, name: $name, type: $type, '
        'isEarned: $isEarned, progress: $progress/$conditionValue)';
  }
}
